### PR TITLE
Rework extractors

### LINKED ISSUE #5274: Switching with "case null"
(First of all, I have read through the changes in https://github.com/HaxeFoundation/haxe/pull/4940)

```haxe
class Main {
	public static function main() {
		switch getString() {
			case null: trace('null');
			case _.split(",") => [s1, s2]: trace(s1, s2);
			default: trace('default');
		}
	}
	
	static function getString():String return null;
}
```

This code fails with `null has no field split`, because the pattern `_.split(',')` is eagerly evaluated. Is it possible to treat null as a special case such that it does a null check first, before evaluating the following statements?

### LINKED ISSUE #11062: Switch with at least one field match accesses all the fields.
This piece of code [traces](https://try.haxe.org/#c8246Dbf) `called` even though nothing access `bar`:

```haxe
class Test {
    static function main() {
        var foo = new Foo();
        
        switch foo {
            case { foo: 123 }:
                trace('yes!');
            case { foo: 321 }:
                trace('yes!');
            default:
        }
    }
}

class Foo {
    public var foo:Int = 5;
    
    public var bar(get, never):Int;
    function get_bar() {
        trace('called');
        return 4;
    }
    
    public function new() {}
}
```

This broke between 3.4.7 and 4.0.5. So in any case, not a regression (within Haxe 4).

It is quite unfortunate for getters that cause actual side effects (like tracking access), or even just do relatively expensive computations (in the worst case synchronized via a lock/mutex). Also potentially sabotages DCE. 

I would speculate that maybe this got introduced in Haxe 4, because as the analyzer got better at cleaning up things, the compiler got a bit more liberal about churning out temp stuff. Unfortunately, non-final methods are never pure, so the getter call remains (even if I remove the trace). In my head, the intermediary AST has something like `var _tmp_bar = foo.get_bar();` (I can see the temp var for the `foo` field, so maybe I'm not too far off) and then the analyzer deletes the var, but not the init expr, because that's not pure. 

If that is the problem, then one solution would be to support `@:pure` on expression level (could be a useful footgun for macros too) and then generate something like `var _tmp_bar = @:pure foo.get_bar();`.

### LINKED ISSUE #11127: Review extractor implementation
While looking into #11062 I came across my extractor implementation, and quite frankly I have no idea wtf is going on there. It's not even a lot of code (less than 50 lines) and it has been working quite well, but it definitely needs some review and documentation in order to be maintainable. 

One related issue is #5274 which required a bit of a workaround.

### PR DESCRIPTION
**Pull request by Simn**
Now the implementation looks like something humans might be able to understand. I checked unit.js, the only difference is some temp variable names. 

Closes #11127
Closes #11062
Closes #5274 again in a nicer way

### CODE REVIEW

File: tests/unit/src/unit/TestMatch.hx [hunk 1]
```diff
@@ -304,109 +308,112 @@ class TestMatch extends Test {
...
 		eq(r, 3);
 
-		var e1 = macro if (1) 2;
-		var e2 = macro if (1) 2 else 3;
+		var e1 = macro if (1)
+			2;
+		var e2 = macro if (1)
+			2
+		else
+			3;
```
Comment (kLabz) [inline tests/unit/src/unit/TestMatch.hx:414]: Is this some haxe formatter bug? @AlexHaxe (especially the first line, that isn't displayed in preview)

Comment (Simn) [reply tests/unit/src/unit/TestMatch.hx:414]: I should really pay more attention when saving these older files...

Comment (kLabz) [reply tests/unit/src/unit/TestMatch.hx:414]: Well it's a lot of noise but if we have a formatting config that fits what we want for haxe code it shouldn't matter much eventually

Comment (AlexHaxe) [reply tests/unit/src/unit/TestMatch.hx:414]: yes, that's a bug, it should've stayed in one line, but `macro` keyword threw it off, treating it like a standalone if.
should be fixed soon.

**Pull request merged**